compiler/rustc_typeck/src/check/check.rs

```diff
@@ -1327,28 +1329,53 @@ pub(super) fn check_transparent<'tcx>(tcx: TyCtxt<'tcx>, sp: Span, adt: ty::AdtD
         let layout = tcx.layout_of(param_env.and(ty));
         // We are currently checking the type this field came from, so it must be local
         let span = tcx.hir().span_if_local(field.did).unwrap();
-        let zst = layout.map_or(false, |layout| layout.is_zst());
-        let align1 = layout.map_or(false, |layout| layout.align.abi.bytes() == 1);
-        (span, zst, align1)
+        let array_len = match ty.kind() {
+            Array(_, len) => len.try_eval_usize(tcx, param_env),
+            _ => None,
+        };
+        let zst = array_len == Some(0) || layout.map_or(false, |layout| layout.is_zst());
```

Comment:
Wait, doesn't this change the actual typechecking behavior?

Reply:
well, i guess not necessarily, because the align check below will still fail? 

I would rather move this is-array check to line 1348, like `if (zst || zero_sized_array) && align != Some(1)` so we don't need to edit this logic, just in case?

Reply:
I've split `zero_sized_array` from `zst`, but the logic stays the same, otherwise we would get #98064 again.

Reply:
Also with a more complex zero-sized but non-trivially-aligned field, we still get the same wrong diagnostic:
```rust
struct ComplexZst<A, B> {
    a: [A; 0],
    b: [B; 0],
}

#[repr(transparent)]
struct Aligned<T, A, B> {
    align: ComplexZst<A, B>,
    value: T,
}
```

```
error[E0690]: transparent struct needs at most one non-zero-sized field, but has 2
 --> src/lib.rs:7:1
  |
7 | struct Aligned<T, A, B> {
  | ^^^^^^^^^^^^^^^^^^^^^^^ needs at most one non-zero-sized field, but has 2
8 |     align: ComplexZst<A, B>,
  |     ----------------------- this field is non-zero-sized
9 |     value: T,
  |     -------- this field is non-zero-sized

For more information about this error, try `rustc --explain E0690`.
```

Is it worth trying to fix this? Should there be something like `ty.is_zst(tcx, param_env)`? I am not really experienced with rustc internals.